crypto/fipsmodule/rand/fork_detect.c

```diff
@@ -42,37 +42,108 @@ DEFINE_STATIC_MUTEX(g_fork_detect_lock)
 DEFINE_BSS_GET(volatile char *, g_fork_detect_addr)
 DEFINE_BSS_GET(uint64_t, g_fork_generation)
 DEFINE_BSS_GET(int, g_ignore_madv_wipeonfork)
+DEFINE_BSS_GET(int, g_ignore_pthread_atfork)
 
+static int init_fork_detect_madv_wipeonfork(void *addr, long page_size) {
+
+  // Some versions of qemu (up to at least 5.0.0-rc4, see linux-user/syscall.c)
+  // ignore |madvise| calls and just return zero (i.e. success). But we need to
+  // know whether MADV_WIPEONFORK actually took effect. Therefore try an invalid
+  // call to check that the implementation of |madvise| is actually rejecting
+  // unknown |advice| values.
+  if (madvise(addr, (size_t)page_size, -1) == 0 ||
+      madvise(addr, (size_t)page_size, MADV_WIPEONFORK) != 0) {
+    munmap(addr, (size_t)page_size);
+    return 0;
+  }
+
+  return 1;
+}
+
+static void pthread_atfork_on_fork(void) {
+
+  struct CRYPTO_STATIC_MUTEX *const lock = g_fork_detect_lock_bss_get();
+
+  // This zeroises the first byte of the memory page pointed to by
+  // |*g_fork_detect_addr_bss_get|. This is the same byte used as fork
+  // detection sentinel in |CRYPTO_get_fork_generation|. The same memory page,
+  // and in turn, the byte, is also the memory zeroised by the |MADV_WIPEONFORK|
+  // fork detection mechanism.
+  //
+  // Aquire locks to be on the safe side. We want to avoid the checks in
+  // |CRYPTO_get_fork_generation| getting executed before setting the sentinel
+  // flag. The write lock prevents any other thread from owning any other type
+  // of lock.
+  CRYPTO_STATIC_MUTEX_lock_write(lock);
+  volatile char *const flag_ptr = *g_fork_detect_addr_bss_get();
```

Comment:
shouldn't this be `volatile const char *flag_ptr`? it's the first time I see `const` after the type and the star :)

Reply:
We modify the memory below, so the memory cannot be declared constant. But the pointer can, and we do not want the pointer to change.